### ISSUE TITLE
GooglePerfKitBenchmarker :FIO hangs in " in __lll_lock_wait () from /lib64/libpthread.so.0"

### ISSUE THREAD

**Issue by srinivaslpalli** [GoogleCloudPlatform/PerfKitBenchmarker#499]
I am trying to run FIO from a PerfKitBenchmarker-0.17.0 installed system simultaneously on multiple VMs
(>16VMs) using different json files (having different VM IP addresses) and see FIO hangs on some of the VMs.

I have included below stack details from both the perfkit client from where I am running pkb.py and on the VM stack
from one of the hung FIO runs.. similar stack seen on all hung FIO runs on different VMs ..
From Perfkit client

[root@scan01cn26 PerfKitBenchmarker-0.17.0]# /usr/local/bin/python ./pkb.py --static_vm_file=VM17_xvdc.json --os_type=rhel --benchmarks=fio --against_device=True --ip_addresses=EXTERNAL
..
2015-09-27 21:01:51,880 156bdfa6 MainThread fio(1/1) INFO FIO prepare on 10.252.136.129
2015-09-27 21:01:51,881 156bdfa6 MainThread fio(1/1) INFO Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o P referredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo yum install -y libaio-devel libaio bc
.
2015-09-27 21:01:53,436 156bdfa6 MainThread fio(1/1) INFO Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o P referredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo yum groupinstall -y "Development Tools"
.
2015-09-27 21:01:54,998 156bdfa6 MainThread fio(1/1) INFO Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o P referredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key git clone https://git.kernel.org/pub/scm/linux/kernel/git/axboe/fio.git /tmp/pkb/fio
.
2015-09-27 21:02:01,884 156bdfa6 MainThread fio(1/1) INFO Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o P referredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key cd /tmp/pkb/fio && git checkout fio-2.1.14
.
2015-09-27 21:02:02,333 156bdfa6 MainThread fio(1/1) INFO Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o P referredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key cd /tmp/pkb/fio && ./configure && make
...
2015-09-27 21:06:35,437 156bdfa6 MainThread fio(1/1) INFO Running: scp -P 22 -pr -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o S erverAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key /tmp/perfkitbenchmarker/run_156bdfa6/fio.job opc@10.252.136.129:
2015-09-27 21:06:35,989 156bdfa6 MainThread fio(1/1) INFO Running benchmark fio
2015-09-27 21:06:35,989 156bdfa6 MainThread fio(1/1) INFO FIO running on 10.252.136.129
2015-09-27 21:06:35,989 156bdfa6 MainThread fio(1/1) INFO FIO Results:
2015-09-27 21:06:35,989 156bdfa6 MainThread fio(1/1) INFO Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o P referredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo /tmp/pkb/fio/fio --output-format=json --filename=/dev/xvdc fio.job

==> hangs here

root@scan01cn26 PerfKitBenchmarker-0.17.0]# ps -ef | grep 10.252.136.129
root 1894 26205 0 00:42 pts/47 00:00:00 grep 10.252.136.129
root 16576 13384 0 Sep27 pts/17 00:00:00 ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo /tmp/pkb/fio/fio --output-format=json --filename=/dev/xvdc fio.job

[root@scan01cn26 PerfKitBenchmarker-0.17.0]# pstack 13384
Thread 2 (Thread 0x7f596f90d940 (LWP 16577)):
#0 0x0000003f4700ce51 in sem_wait () from /lib64/libpthread.so.0
#1 0x00000000004ce7bd in PyThread_acquire_lock ()
#2 0x00000000004d1760 in lock_PyThread_acquire_lock ()
#3 0x000000000049a3cd in PyEval_EvalFrameEx ()
#4 0x000000000049bb5b in PyEval_EvalCodeEx ()
#5 0x0000000000499549 in PyEval_EvalFrameEx ()
#6 0x000000000049bb5b in PyEval_EvalCodeEx ()
#7 0x0000000000499549 in PyEval_EvalFrameEx ()
#8 0x000000000049a619 in PyEval_EvalFrameEx ()
#9 0x000000000049a619 in PyEval_EvalFrameEx ()
#10 0x000000000049bb5b in PyEval_EvalCodeEx ()
#11 0x00000000004fc4cf in function_call ()
#12 0x0000000000418ecd in PyObject_Call ()
#13 0x00000000004203ff in instancemethod_call ()
#14 0x0000000000418ecd in PyObject_Call ()
#15 0x0000000000492a76 in PyEval_CallObjectWithKeywords ()
#16 0x00000000004d1c22 in t_bootstrap ()
#17 0x0000003f4700683d in start_thread () from /lib64/libpthread.so.0
#18 0x0000003f460d511d in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f59747746e0 (LWP 13384)):
#0 0x0000003f460cc0a6 in poll () from /lib64/libc.so.6
#1 0x00007f5970efc89b in poll_poll () from /usr/local/lib/python2.7/lib-dynload/select.so
#2 0x000000000049a3cd in PyEval_EvalFrameEx ()
#3 0x000000000049bb5b in PyEval_EvalCodeEx ()
#4 0x0000000000499549 in PyEval_EvalFrameEx ()
#5 0x000000000049a619 in PyEval_EvalFrameEx ()
#6 0x000000000049bb5b in PyEval_EvalCodeEx ()
#7 0x0000000000499549 in PyEval_EvalFrameEx ()
#8 0x000000000049bb5b in PyEval_EvalCodeEx ()
#9 0x0000000000499549 in PyEval_EvalFrameEx ()
#10 0x000000000049bb5b in PyEval_EvalCodeEx ()
#11 0x0000000000499549 in PyEval_EvalFrameEx ()
#12 0x000000000049a619 in PyEval_EvalFrameEx ()
#13 0x000000000049a619 in PyEval_EvalFrameEx ()
#14 0x000000000049a619 in PyEval_EvalFrameEx ()
#15 0x000000000049bb5b in PyEval_EvalCodeEx ()
#16 0x0000000000499549 in PyEval_EvalFrameEx ()
#17 0x000000000049bb5b in PyEval_EvalCodeEx ()
#18 0x0000000000499549 in PyEval_EvalFrameEx ()
#19 0x000000000049bb5b in PyEval_EvalCodeEx ()
#20 0x000000000049bbd2 in PyEval_EvalCode ()
#21 0x00000000004bdca1 in PyRun_FileExFlags ()
#22 0x00000000004bdf3a in PyRun_SimpleFileExFlags ()
#23 0x0000000000414f85 in Py_Main ()
#24 0x0000003f4601d9f4 in __libc_start_main () from /lib64/libc.so.6
#25 0x0000000000414119 in _start ()

[root@scan01cn26 PerfKitBenchmarker-0.17.0]# pstack 16576
#0 0x00007fb3c99e0113 in __select_nocancel () from /lib64/libc.so.6
#1 0x00007fb3cba824e5 in ?? ()
#2 0x00007fb3cba84010 in ?? ()
#3 0x00007fb3cba7e311 in main ()
On the VM :

[root@fc74b1 opc]# ps -ef | grep pkb
root 6832 1 0 17:09 ? 00:00:50 /tmp/pkb/fio/fio --output-format=json --filename=/dev/xvdc fio.job
root 7389 7373 0 20:28 pts/0 00:00:00 grep pkb

[root@fc74b1 opc]# pstack 6832
#0 0x0000003696a0e264 in __lll_lock_wait () from /lib64/libpthread.so.0
#1 0x0000003696a0b69f in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#2 0x000000000042f699 in fio_mutex_down ()
#3 0x000000000044a873 in thread_main ()
#4 0x000000000044bacc in run_threads ()
#5 0x000000000044bebb in fio_backend ()
#6 0x000000369661ed5d in __libc_start_main () from /lib64/libc.so.6
#7 0x0000000000409e59 in _start ()

Not sure if this is a known issue, any pointers, feedback is appreciated.

Thanks

**Comment by yuyantingzero** [comment]
Hi @srinivaslpalli , could you provide some details on your guest vm (how much memory, disk size, kernel version)? Also, how often does this occur?

**Comment by yuyantingzero** [comment]
I noticed you are using against_device=True but also using static vms, which might be the problem. Since static vms doesn't support device path: https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/341. Could you provide a sample static vm file you are using and the fio.job file on your guest vm (fio benchmark will modify job file if against_device flag is used)?

**Comment by rmathara** [comment]
Hi , answering on behalf of my colleague srinivas,

- Yes, we are aware of that issue #341, fio_benchmark.py is hardcoded to use raw device and it works fine if we run the test individually one at a time on these VMs (identical config, uek3 kernel, 
4 core,30GB) only when we run FIO on these VMs simultaneously, at random FIO hangs on different VMs with above stack. Say we run on 20 VMs in parallel, FIO hangs on VM7, VM 13,
repeat again, hangs on VM 5 and VM 18.. doesn't happen to the same VM everytime. 

I see this issue https://github.com/axboe/fio/issues/52 , not sure if we are hitting this issue, based on the stack we are seeing. 

fio.job 
..
[global]
ioengine=libaio


[random_write_test]
stonewall
overwrite=1
rw=randwrite
blocksize=8k
direct=1
iodepth=4096
filename=/dev/xvdc
runtime=180
time_based=1

[random_read_test]
stonewall
rw=randread
blocksize=8k
direct=1
iodepth=512
filename=/dev/xvdc
runtime=180
time_based=1
....

Thanks

**Comment by yuyantingzero** [comment]
@rmathara , what do you mean by simultaneously? I do not think fio benchmark involves any communications between vms, it should make no difference, could you provide exact command you use to run simultaneously? Basically, I wonder if you are using some shell script to start multiple pkb runs or you are using parallelism flag?

Also, I noticed https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/472, I would like rule out the possibility that your walk around did anything wrong. Could you attach log file of the failed run? (can be found in /tmp/perfkitbenchmarker/run_xxx/pkb.log)

**Comment by rmathara** [comment]
@srinivaslpalli provided that earlier .. 

Here it is - We run below cmd on 24 different terminals using 24 different json files (which point to respective VM ip address) .. below example is for VM17. No shell scripts involved.. just executing on all 24 terminals below cmd with a different static_vm_file 
..
# /usr/local/bin/python ./pkb.py --static_vm_file=VM17_xvdc.json --os_type=rhel --benchmarks=fio --against_device=True --ip_addresses=EXTERNAL
....

cat VM17_xvdc.json
[
 {
  "ip_address": "10.252.136.129",
  "user_name": "opc",
  "keyfile_path": "/root/reddy/ssh_key" ,
  "os_type":"rhel",
  "internal_ip": "10.252.136.129",
  "scratch_disk_mountpoints": ["/dev/xvdc"]
 }
]

Run output:
....
[root@scan01cn26 PerfKitBenchmarker-0.17.0]# /usr/local/bin/python ./pkb.py --static_vm_file=VM17_xvdc.json --os_type=rhel --benchmarks=fio --against_device=True --ip_addresses=EXTERNAL
2015-09-27 21:01:49,295 156bdfa6 MainThread INFO     Verbose logging to: /tmp/perfkitbenchmarker/run_156bdfa6/pkb.log
2015-09-27 21:01:49,296 156bdfa6 MainThread INFO     Flag values:
--ip_addresses=EXTERNAL
--os_type=rhel
--static_vm_file=VM17_xvdc.json
--against_device
--benchmarks=fio
2015-09-27 21:01:49,302 156bdfa6 MainThread fio(1/1) INFO     Preparing benchmark fio
2015-09-27 21:01:49,304 156bdfa6 Thread-1 fio(1/1) INFO     VM: 10.252.136.129
2015-09-27 21:01:49,304 156bdfa6 Thread-1 fio(1/1) INFO     Waiting for boot completion.
2015-09-27 21:01:49,305 156bdfa6 Thread-1 fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key hostname2015-09-27 21:01:49,721 156bdfa6 Thread-1 fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key -t -t bash -l -c "echo 'Defaults:opc !requiretty' | sudo tee /etc/sudoers.d/pkb"
2015-09-27 21:01:50,200 156bdfa6 Thread-1 fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key mkdir -p/tmp/pkb
2015-09-27 21:01:50,629 156bdfa6 Thread-1 fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key rpm -qa > /tmp/pkb/rpm_package_list
2015-09-27 21:01:51,874 156bdfa6 MainThread fio(1/1) WARNING  File resource loader root data isnot a directory.
2015-09-27 21:01:51,879 156bdfa6 MainThread fio(1/1) INFO     ssh to VMs in this benchmark by na me with: ssh -F /tmp/perfkitbenchmarker/run_156bdfa6/ssh_config <vm name>
2015-09-27 21:01:51,880 156bdfa6 MainThread fio(1/1) INFO     FIO prepare on 10.252.136.129
2015-09-27 21:01:51,881 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo yu m install -y libaio-devel libaio bc
2015-09-27 21:01:53,436 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo yu m groupinstall -y "Development Tools"
2015-09-27 21:01:54,998 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key git clone https://git.kernel.org/pub/scm/linux/kernel/git/axboe/fio.git /tmp/pkb/fio
2015-09-27 21:02:01,884 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key cd /tmp/pkb/fio && git checkout fio-2.1.14
2015-09-27 21:02:02,333 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key cd /tmp/pkb/fio && ./configure && make
2015-09-27 21:02:24,448 156bdfa6 MainThread fio(1/1) WARNING  File resource loader root data isnot a directory.
2015-09-27 21:02:24,448 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key df -k -P /dev/xvdc | tail -n +2 | awk '{ print $2 }'
2015-09-27 21:02:24,898 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key cat /proc/meminfo | grep MemTotal | awk '{print $2}'
/dev/xvdc
2015-09-27 21:02:25,361 156bdfa6 MainThread fio(1/1) INFO     Fill scratch disk on 10.252.136.129 at /dev/xvdc
2015-09-27 21:02:25,361 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo /t mp/pkb/fio/fio --filename=/dev/xvdc --ioengine=libaio --name=fill-device --blocksize=512k --iodepth=64 --rw=write --direct=1 --size=100%
2015-09-27 21:06:35,436 156bdfa6 MainThread fio(1/1) INFO     Removing directory and filename in  job file.
2015-09-27 21:06:35,436 156bdfa6 MainThread fio(1/1) WARNING  File resource loader root data isnot a directory.
2015-09-27 21:06:35,437 156bdfa6 MainThread fio(1/1) INFO     Write modified job file to temp di rectory /tmp/perfkitbenchmarker/run_156bdfa6
2015-09-27 21:06:35,437 156bdfa6 MainThread fio(1/1) INFO     Running: scp -P 22 -pr -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key /tmp/perfkitbenchmarker/run_156bdfa6/fio.job opc@10.252.136.129:
2015-09-27 21:06:35,989 156bdfa6 MainThread fio(1/1) INFO     Running benchmark fio
2015-09-27 21:06:35,989 156bdfa6 MainThread fio(1/1) INFO     FIO running on 10.252.136.129
2015-09-27 21:06:35,989 156bdfa6 MainThread fio(1/1) INFO     FIO Results:
2015-09-27 21:06:35,989 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo /t mp/pkb/fio/fio --output-format=json --filename=/dev/xvdc fio.job
=> hangs here

Thanks

**Comment by yuyantingzero** [comment]
They are using v0.17.0 version of PKB, which has nothing to do with noahl's recent refactoring. Probabily related to https://github.com/axboe/fio/issues/52. I am still investigating.

**Comment by yuyantingzero** [comment]
Hi @rmathara , I took a close look at the logs you send me, everything looks fine. It is very likely caused by fio binary itself.

I haven't been able to repro the problem you are seeing. One thing you could try is run the fio command manually with --debug=all,
Basically try, sudo /tmp/pkb/fio/fio --debug=all --filename=/dev/xvdc fio.job > debug.
To make sure this is a problem in fio, would you mind opening an issue at fio repository to have them take a look.
Also, you mentioned the benchmark runs fine sequentially. How many sequential run have you tried? Sounds like when you run simultaneously, there is 10% repro rate. I would be surprised if you run 20 times sequentially without any repro. Would you be able to provide more details on how you setup your static vms (is it from cloud or this is a physical machine)?

**Comment by stfeng** [comment]
@rmathara, I just noticed that the version of PKB you are using is 0.17.0. It might be worthwhile to upgrade it to the latest version 0.21.0 and see what happens. There used to be a bug in FIO benchmark where in directly-against-device mode, we didn't umount the file system first before we start the test. This has been fixed with recent releases.

**Comment by rmathara** [comment]
@yuyantingzero, Thanks for your time and taking a look,  The issue I have in running  fio on the VM is it gets removed per below pkb.log :

2015-09-27 21:06:35,989 156bdfa6 MainThread fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo /t mp/pkb/fio/fio --output-format=json --filename=/dev/xvdc fio.job

==> after it hangs here , after waiting for some time, I  ctrl-C out of this
^C
2015-09-28 00:44:19,421 156bdfa6 Thread-16 fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -oPreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key rpm -q a | grep --fixed-strings --line-regexp --invert-match --file /tmp/pkb/rpm_package_list | xargs --no-run-if-empty sudo rpm -e

2015-09-28 00:44:20,631 156bdfa6 Thread-16 fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o Pr eferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key rm -rf /tmp/pkb

==> Above is where it does "rm -rf /tmp/pkb" and on the VM though I see this below process 
==> on the VM, I won't have "/tmp/pkb/fio/fio " source to run --debug as you suggested.

root      6832     1  0 17:09 ?        00:00:50 /tmp/pkb/fio/fio --output-format=json --filename=/dev/xvdc fio.jo
..
2015-09-28 00:44:21,054 156bdfa6 Thread-16 fio(1/1) INFO     Running: ssh -A -p 22 opc@10.252.136.129 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o Pr eferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /root/reddy/ssh_key sudo rm/etc/sudoers.d/pkb
2015-09-28 00:44:21,490 156bdfa6 MainThread fio(1/1) INFO     Complete logs can be found at: /tmp/perfkitbenchmarker/run_156bdfa6/pkb.log

Question :

Anyways I can include the --debug flag in the fio.job or fio_benchmark.py or as part of this below pkb cmd line ?
..
usr/local/bin/python ./pkb.py --static_vm_file=VM17_xvdc.json --os_type=rhel --benchmarks=fio --against_device=True --ip_addresses=EXTERNAL
...

Answers to your questions
 - Yes, I will open this issue at the fio repository
-  No issues running sequentially.. tested over 50 times on these VMs no issues
- these VMs provisioned from Cloud

@stfeng, our tests are run using raw devices, would like to get some details if/how this could be impacting us  in light of your below comments?
There used to be a bug in FIO benchmark where in directly-against-device mode, we didn't umount the file system first before we start the test. This has been fixed with recent releases.

Thanks

**Comment by stfeng** [comment]
@rmathara 

If we are testing (RW) directly against raw device, then that device should not be in-use by the file system at the same time. If so, file system metadata/inodes/master block, etc would soon be corrupted by fio writes, and then depending on the OS, the file system would start panicking and possibly hang. Even if the FS does not hang, it's highly likely it might attempt to "repair" the block device by marking blocks as bad and those writes would contend with FIO's writes to those blocks and possibly causing hang at both places.(FIO and/or file system).

In any rate, it's just better to umount the device first before we use the raw device as the target of the FIO test.

**Comment by yuyantingzero** [comment]
@rmathara Thanks for the detailed response. Adding --debug flag into pkb will break the result parsing, there is no easy way to do this. Sorry about that. 

But to run manually, you can do the following:
./pkb.py --static_vm_file=VM17_xvdc.json --os_type=rhel --benchmarks=fio --against_device=True --ip_addresses=EXTERNAL --run_stage=prepare
The above step will install fio and dependency you need and give you a run_uri. And then, you can login to the vm and run the debug command manually.

To run the benchmark as it is (you can repeat the run stage as many time as you want):
./pkb.py --static_vm_file=VM17_xvdc.json --os_type=rhel --benchmarks=fio --against_device=True --ip_addresses=EXTERNAL --run_stage=run --run_uri=RUN_URI
To cleanup the packages:
./pkb.py --static_vm_file=VM17_xvdc.json --os_type=rhel --benchmarks=fio --against_device=True --ip_addresses=EXTERNAL --run_stage=cleanup --run_uri=RUN_URI
These break down stages also save you the time to prewarm the disks if you are reusing the same cluster of vms.

Also, from your description, your simultaneous run actually looks like 24 individual runs at the same time, should not make any difference. It is not very likely but could you make a quick check to make sure every static json file points to a different vm?
"cat *_xvdc.json | grep internal_ip | uniq | wc -l" will do.

**Comment by rmathara** [comment]
Hi: 

our internal IPs are all unique. 
Here is the update I get from FIO repository for my issue and is closed now..
..
This is a duplicate of issue #52, as you suspected. They need to update their fio version, or pull that fix into their base.
..
any next steps, suggestions for me ?

Thanks

**Comment by yuyantingzero** [comment]
Opened a new issue #504 to update fio package. I will come up with a pull request today and let you know when it is merged.
Closing this issue for now. Feel free to reopen if you have more question/concerns.
